### ISSUE TITLE
[A11Y] Moving blocks within layouts does not work for people using screen readers

### ISSUE THREAD

**Issue by indigoxela** [backdrop/backdrop-issues#6326]
## Description of the bug

This forum comment clearly describes, why our current approach to provide the ability to rearrange blocks without mouse is far from ideal:
https://forum.backdropcms.org/forum/how-do-i-define-weight-block-layout?page=0%2C0#comment-5930

Accessibility matters!

## Steps To Reproduce

Obviously a screen reader is required to reproduce the problem.

## Actual behavior

It's very hard, if not impossible to administer layouts when using a screen reader. But it shouldn't. At least switching from arrow keys to something else seems necessary.

## Additional information

We tried to improve keyboard navigation in https://github.com/backdrop/backdrop-issues/issues/4941, but obviously, we're not there, yet.

**Comment by jenlampton** [comment]
There was one relevant example provided by @klonos  in the other issue, I'd like to add here too for consideration:

Also: [sitepoint.com/accessible-drag-drop](https://www.sitepoint.com/accessible-drag-drop) ...with the code from that tutorial available here: [sitepoint-editors/accessible-drag-and-drop](https://github.com/sitepoint-editors/accessible-drag-and-drop)

It sounds like the biggest missing piece is moving a block to a different region?

So far suggestions for that have included:
* Adding a region selection option to the configure blocks page (not ideal)
* Add an visually hidden region selection option to each block on the manage blocks page
* Find a widely accepted solution and adopt whatever that is :)

**Comment by indigoxela** [comment]
Hi @suriyan2538, and welcome in our issue queue.

Can you describe a bit, what the ComboBox you mentioned is? I'm having trouble to understand your workflow to move blocks to different regions.

And I'd also like to know, which screen reader product you use, and which keys are reserved by it. All arrow keys seem to be used to navigate on the page. Are there additional ones, that we should avoid to use, to prevent conflicts or misunderstandings?

**Comment by indigoxela** [comment]
@suriyan2538 many thanks for your helpful explanations!

Wow, I wasn't aware of that additional form item, which is only available when calling the block form directly. So, we already seem to have what we need for switching regions directly, but for some reason it's only available via obscure methods.

@quicksketch I wonder, why this isn't available by default, visually hidden, but available for screen reader users. Here's [a link to the code, last edited 2016](https://github.com/backdrop/backdrop/blob/1.x/core/modules/layout/layout.admin.inc#L1583).

Now for a better keyboard approach for moving a block within a region.

I'm not sure if we can provide a combination of arrow and enter key. What about a single key, like "f" for forward and "b" for back? Just as an example. Could that work for you @suriyan2538?

Currently we add aria-labels to the individual blocks on the page, is that label read aloud? I'm aware that on that form there are lots of items with a tab index assigned, so I'm not sure if that isn't already quite confusing.

**Comment by suriyan2538** [comment]
This is what screen readers can read.

Skip to main content
Header
Link• Add block
Link• Configure region
Header block
Link• Configure
Link• Remove
Link• Disable
Contains elements typical of a site header including the site logo, name, slogan, and optionally a navigational menu.
Primary navigation
Link• Configure
Link• Remove
Link• Disable
A list of links for the menu: Primary navigation.

When I set up my Block, I do so through these links.

For the most part, it works fine. There is only block arrangement. that cannot be accessed by screen readers

Personally, I think Even if we change the scroll up and down keys. But screen readers likely won't be able to access it.

Because we can't set the screen reader to read aloud. When moving blocks to different positions on the website page

**Comment by suriyan2538** [comment]
@indigoxela

You have completely summarized the issue.

**Comment by docwilmot** [comment]
For the record, after tabbing to the block, moving blocks between regions works with the Right and Left arrows, while moving blocks within regions works with the Up and Down arrows. I suspect that may not be the problem here though? I admit its not quite clear to me what needs fixing.

**Comment by indigoxela** [comment]
> I admit its not quite clear to me what needs fixing.

Hey, @docwilmot, many thanks for chiming in. Give the summary in [this comment](https://github.com/backdrop/backdrop-issues/issues/6326#issuecomment-1847284440) a try.

**Comment by docwilmot** [comment]
I saw that, but its not clear why arrows keys arent the answer.

**Comment by klonos** [comment]
> Wow, I wasn't aware of that additional form item, which is only available when calling the block form directly. ...

@indigoxela that is by intention I believe. You can get to that form either if:
- JS is disabled on your browser, in which case clicking the "configure" link will take you to that form instead of opening a dialog (which relies on JS)
- you intentionally right-click the "configure" link and select "open in new tab", in which case we need to show the region selection menu, as the user has no longer any region context available within their viewport.

I hope that this helps clarify why things work that way. It seems that we didn't do too much testing with screen readers when we initially built this UI.

**Comment by suriyan2538** [comment]
@indigoxela
@klonos

> In both cases we need to show the region selection menu, as the user has moved away from the main "Manage blocks" UI and doesn't have any region context available within their viewport any longer.
> 

Will it be possible? We will put the block arrangement on this page. without having to mess with the main layout

Because for screen reader users It wasn't difficult at all. Because we don't need to see the live block arrangement. As for placing blocks on the left or right side You can use the selection menu instead of paste. Using keyboard keys to arrange blocks Probably too difficult for screen readers.

I think that Arranging blocks using the on web menu Probably most appropriate for screen readers.
